extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/TenantConfigBean.java

```diff
@@ -1,46 +1,85 @@
 package io.quarkus.oidc.runtime;
 
+import static java.util.Collections.unmodifiableMap;
+
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Function;
 
 import jakarta.enterprise.context.spi.CreationalContext;
 
+import org.jboss.logging.Logger;
+
 import io.quarkus.arc.BeanDestroyer;
 import io.quarkus.oidc.OidcTenantConfig;
+import io.quarkus.runtime.configuration.ConfigurationException;
 import io.smallrye.mutiny.Uni;
 
 public class TenantConfigBean {
 
+    private static final Logger LOG = Logger.getLogger(TenantConfigBean.class);
+
     private final Map<String, TenantConfigContext> staticTenantsConfig;
     private final Map<String, TenantConfigContext> dynamicTenantsConfig;
     private final TenantConfigContext defaultTenant;
-    private final Function<OidcTenantConfig, Uni<TenantConfigContext>> tenantConfigContextFactory;
+    private final TenantContextFactory tenantContextFactory;
+
+    @FunctionalInterface
+    public interface TenantContextFactory {
+        Uni<TenantConfigContext> create(OidcTenantConfig oidcTenantConfig, boolean dynamicTenant, String tenantId);
+    }
 
     public TenantConfigBean(
             Map<String, TenantConfigContext> staticTenantsConfig,
-            Map<String, TenantConfigContext> dynamicTenantsConfig,
             TenantConfigContext defaultTenant,
-            Function<OidcTenantConfig, Uni<TenantConfigContext>> tenantConfigContextFactory) {
-        this.staticTenantsConfig = staticTenantsConfig;
-        this.dynamicTenantsConfig = dynamicTenantsConfig;
+            TenantContextFactory tenantContextFactory) {
+        this.staticTenantsConfig = new ConcurrentHashMap<>(staticTenantsConfig);
+        this.dynamicTenantsConfig = new ConcurrentHashMap<>();
         this.defaultTenant = defaultTenant;
-        this.tenantConfigContextFactory = tenantConfigContextFactory;
+        this.tenantContextFactory = tenantContextFactory;
     }
 
-    public Map<String, TenantConfigContext> getStaticTenantsConfig() {
-        return staticTenantsConfig;
+    public Uni<TenantConfigContext> createTenantContext(OidcTenantConfig oidcConfig, boolean dynamicTenant) {
+        var tenantId = oidcConfig.getTenantId().orElseThrow();
+        if (dynamicTenant && oidcConfig.logout.backchannel.path.isPresent()) {
+            throw new ConfigurationException(
+                    "BackChannel Logout is currently not supported for dynamic tenants (tenant ID: " + tenantId + ")");
+        }
+        var tenants = dynamicTenant ? dynamicTenantsConfig : staticTenantsConfig;
```

Comment:
I'm not sure we even need it, since static contexts are already in the static map, right ? So I guess we need to keep the static map as non-concurrent after all :-), just avoid adding an extra reference to the initialized one in the dynamic map

Reply:
Not all. Looks like the `Default` tenant is not always in the map - at least a bunch of tests in `integration-tests/oidc-wiremock/` fail with that assumption.
We also need the map to be an CHM - it can be modified, if the `Default` tenant is added (not sure why it can be missing, but it can be) or a non-ready tenant is created.

I think it's possible to optimize this - and actually also resolving the "concurrent tenant creation" issue above.

Reply:
`TenantConfigContext` is immutable, so this replaces original not ready instance with a new one. Problem is that whoever kept reference to this original `TenantConfigContext` will always have not ready instance and need to keep polling `TenantConfigBean` for changes. #43433 has this map immutable.

Reply:
> So I guess we need to keep the static map as non-concurrent after all 

this is what my previous comment was answering

Reply:
> We also need the map to be an CHM - it can be modified, if the Default tenant is added (not sure why it can be missing, but it can be) or a non-ready tenant is created.

Default tenant will not be available if OIDC server is not available (for example) and it will recover when OIDC server becomes available.